### PR TITLE
FAT-1017: When an overdue item is returned, an overdue fine is billed per the Overdue Fine Policy

### PR DESCRIPTION
**Pull request by yusuf-murodov-epam**
## Purpose
[FAT-1017](https://issues.folio.org/browse/FAT-1017)
_Create Karate test for Loans: When an overdue item is returned, an overdue fine is billed per the Overdue Fine Policy_

## Approach
- Reproduced test scenario with UI.
- Learned necessary sent requests and then necessary received responses 
- Based on the request/response created karate test scenario

## Learning
- learn karate framework from original doc file on the github
- looked other scenarios how they are described.

### CODE REVIEW

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature [hunk 1]
```diff
@@ -630,3 +633,70 @@ Feature: Loans tests
...
+    And match checkOutResponse.response.dueDate contains extDueDate
+
+    # checkIn the item
+    * def postCheckInResponse = call read('classpath:vega/mod-circulation/features/util/initData.feature@CheckInItem') { itemBarcode: #(extItemBarcode), extCheckInDate: #(extCheckInDate) }
+    And match postCheckInResponse.response.loan.id == loanId
+    And match postCheckInResponse.response.loan.status.name == 'Closed'
+    And match postCheckInResponse.response.item.id == itemId
+    And match postCheckInResponse.response.item.status.name == 'Available'
+    And match postCheckInResponse.response.loan.returnDate contains extCheckInDate
+    And match postCheckInResponse.response.loan.dueDate contains extDueDate
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature:678]:
```suggestion
    * call read('classpath:vega/mod-circulation/features/util/initData.feature@CheckInItem') { itemBarcode: #(extItemBarcode), extCheckInDate: #(extCheckInDate) }
```
the same

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature [hunk 2]
```diff
@@ -630,3 +633,70 @@ Feature: Loans tests
...
+    * call read('classpath:vega/mod-circulation/features/util/initData.feature@PostManualCharge')
+    * call read('classpath:vega/mod-circulation/features/util/initData.feature@PostPaymentMethod')
+
+    # checkOut the item
+    * def checkOutResponse = call read('classpath:vega/mod-circulation/features/util/initData.feature@PostCheckOut') { extCheckOutUserBarcode: #(extUserBarcode), extCheckOutItemBarcode: #(extItemBarcode) }
+    * def loanId = checkOutResponse.response.id
+    And match checkOutResponse.response.item.id == itemId
+    And match checkOutResponse.response.item.status.name == 'Checked out'
+    And match checkOutResponse.response.loanDate == '#present'
+    And match checkOutResponse.response.dueDate contains extDueDate
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature:669]:
```suggestion
    * def checkOutResponse = call read('classpath:vega/mod-circulation/features/util/initData.feature@PostCheckOut') { extCheckOutUserBarcode: #(extUserBarcode), extCheckOutItemBarcode: #(extItemBarcode) }
    * def loanId = checkOutResponse.response.id
```
I believe we don't need to verify check-out method in this test

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature [hunk 3]
```diff
@@ -630,3 +633,70 @@ Feature: Loans tests
...
+    # get overdue fine by loan id
+    Given path 'accounts'
+    And param query = 'loanId==' + loanId
+    When method GET
+    Then status 200
+    And match response.totalRecords == 1
+    And match response.accounts[0].status.name == 'Open'
+    And match response.accounts[0].feeFineType == 'Overdue fine'
+    And def accountId = response.accounts[0].id
+    And def overdueFine = response.accounts[0].amount
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature:689]:
```suggestion
    Then status 200
    * def accountInResponse = response.accounts[0]
    And match response.totalRecords == 1
    And match accountInResponse .status.name == 'Open'
    And match accountInResponse .feeFineType == 'Overdue fine'
    And def accountId = accountInResponse .id
    And def overdueFineAmount = accountInResponse .amount
```

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature [hunk 4]
```diff
@@ -630,3 +633,70 @@ Feature: Loans tests
...
+
+    # check-pay and verify that the fine is possible to be paid
+    Given path 'accounts/' + accountId + '/check-pay'
+    And request { amount: '#(overdueFine)' }
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature:693]:
```suggestion
    And request { amount: '#(overdueFineAmount)' }
```

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature [hunk 5]
```diff
@@ -630,3 +633,70 @@ Feature: Loans tests
...
+    When method POST
+    Then status 200
+    And match response.allowed == true
+
+    # pay overdue fine and verify that the fine is billed fully
+    * def postPayResponse = call read('classpath:vega/mod-circulation/features/util/initData.feature@PostPay') { amount: #(overdueFine) }
+    And match postPayResponse.response.feefineactions[0].typeAction == 'Paid fully'
+    And match postPayResponse.response.feefineactions[0].amountAction == overdueFine
+    And match postPayResponse.response.feefineactions[0].accountId == accountId
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature:702]:
```suggestion
    * def postPayResponse = call read('classpath:vega/mod-circulation/features/util/initData.feature@PostPay') { amount: #(overdueFineAmount) }
    * def feefineactionInResponse = postPayResponse.response.feefineactions[0]
    And match feefineactionInResponse .typeAction == 'Paid fully'
    And match feefineactionInResponse .amountAction == overdueFineAmount
    And match feefineactionInResponse .accountId == accountId
```
add empty line at the end

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature [hunk 6]
```diff
@@ -630,3 +633,70 @@ Feature: Loans tests
     And match postCheckInResponse.response.loan.status.name == 'Closed'
     And match postCheckInResponse.response.item.id == itemId
     And match postCheckInResponse.response.item.status.name == 'Available'
+
+  Scenario: When an overdue item is returned, an overdue fine is billed per the Overdue Fine Policy
+
+    * def extItemBarcode = 'FAT-1017IBC'
+    * def extUserBarcode = 'FAT-1017UBC'
+    * def extCheckInDate = '2021-11-17T13:30:46.000Z'
+    * def extDueDate = '2021-11-17T13:25:46'
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature:642]: redundant

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature (hunk 5 above)
Comment (viktor-draban) [reply mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature:702]: In addition add three files ( owner-entity-request.json pay-entity-request.json payment-method-entity-request.json ) under folder "feefine" which is will be located as the same level as folders "instance", "item" etc

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature (hunk 6 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature:642]: could you tell me why they are redundant? if i won't provide itembarcode, then I can't checkout and check in

Comment (viktor-draban) [reply mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature:642]: Could you tell me where did you see any comments about itemBarcode ?

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature (hunk 5 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature]: done

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature (hunk 4 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature]: fixed

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature (hunk 3 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature]: fixed

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature (hunk 2 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature]: fixed

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature (hunk 1 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature]: fixed

File: mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature (hunk 6 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/vega/mod-circulation/features/loans.feature]: you wrote redundant and I saw few lines and thought those vars on those lines were redundant?

File: mod-circulation/src/main/resources/vega/mod-circulation/features/samples/pay-entity-request.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
+  "amount": "#(amount)",
+  "notifyPatron": "false",
+  "paymentMethod": "Payment method",
+  "servicePointId": "#(servicePointId)",
+  "userName": "Name Surname"
+}
```
[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/vega/mod-circulation/features/samples/pay-entity-request.json]: added empty line but after committing the line was cleared away?

File: mod-circulation/src/main/resources/vega/mod-circulation/features/util/initData.feature [hunk 8]
```diff
@@ -298,10 +323,11 @@ Feature: init data for mod-circulation
   Scenario: check in item by barcode
     * def checkInId = call uuid
     * def getDate = read('classpath:vega/mod-circulation/features/util/get-time-now-function.js')
-    * def checkInDate = getDate()
+    * def intCheckInDate = getDate()
```
[the remark this answers is not in the document]
Comment (viktor-draban) [reply mod-circulation/src/main/resources/vega/mod-circulation/features/util/initData.feature:326]: image
Are you sure ?

[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/vega/mod-circulation/features/util/initData.feature:326]: interesting, in my case it failed 8 scenarios

[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/vega/mod-circulation/features/util/initData.feature]: fixed, I have no idea, when i run it yesterday , it failed and now it passed)

**Review by viktor-draban — APPROVED** [review]

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**